File: OrchidCore/src/main/java/com/eden/orchid/api/events/EventServiceImpl.java
package com.eden.orchid.api.events;

import com.caseyjbrooks.clog.Clog;
import com.eden.orchid.api.OrchidContext;
import lombok.Getter;

import javax.inject.Inject;
import javax.inject.Singleton;
import java.lang.reflect.Method;
import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;
import java.util.Stack;

@Singleton
public final class EventServiceImpl implements EventService {

    private OrchidContext context;

    private Set<OrchidEventListener> eventListeners;
    @Getter private Set<EventServiceImpl.EventHandler> eventHandlers;
    private Stack<Class<? extends OrchidEvent>> eventsInProgress;

    static class EventHandler {
        public Class<? extends OrchidEvent> eventClass;
        public boolean allowSubclasses;

        public Object acceptor;
        public Method callback;
    }

    @Inject
    public EventServiceImpl(Set<OrchidEventListener> eventListeners) {
        this.eventHandlers = new HashSet<>();
        this.eventsInProgress = new Stack<>();
        this.eventListeners = eventListeners;
    }

    @Override
    public void initialize(OrchidContext context) {
        this.context = context;
        for (OrchidEventListener listener : eventListeners) {
            registerEventListeners(listener);
        }
    }

    @Override
    public void registerEventListeners(OrchidEventListener listener) {
        Arrays.stream(listener.getClass().getDeclaredMethods())
              .filter(method -> method.isAnnotationPresent(On.class))
              .forEach(method -> {
                  On methodAnnotation = method.getAnnotation(On.class);
                  Class<? extends OrchidEvent> eventType = methodAnnotation.value();

                  EventHandler handler = new EventHandler();
                  handler.eventClass = methodAnnotation.value();
                  handler.allowSubclasses = methodAnnotation.subclasses();
                  handler.acceptor = listener;
                  handler.callback = method;

                  boolean addHandler = true;

                  if(handler.callback.getParameterCount() != 1) {
                      addHandler = false;
                      if(handler.allowSubclasses) {
                          Clog.e("Event handler [{}.{}] must accept a single parameter of type [{}] or a subclass of it.", listener.getClass().getSimpleName(), method.getName(), eventType.getSimpleName());
                      }
                      else {
                          Clog.e("Event handler [{}.{}] must accept a single parameter of type [{}]", listener.getClass().getSimpleName(), method.getName(), eventType.getSimpleName());
                      }
                  }

                  if(handler.allowSubclasses) {
                      if(addHandler && !eventType.isAssignableFrom(handler.callback.getParameterTypes()[0])) {
                          addHandler = false;
                          Clog.e("Event handler [{}.{}] must accept a single parameter of type [{}] or a subclass of it.", listener.getClass().getSimpleName(), method.getName(), eventType.getSimpleName());
                      }
                  }
                  else {
                      if(addHandler && !eventType.equals(handler.callback.getParameterTypes()[0])) {
                          addHandler = false;
                          Clog.e("Event handler [{}.{}] must accept a single parameter of type [{}]", listener.getClass().getSimpleName(), method.getName(), eventType.getSimpleName());
                      }
                  }

                  if(addHandler) {
                      this.eventHandlers.add(handler);
                  }
              });
    }

    @Override
    public void deregisterEventListeners(OrchidEventListener listener) {
        eventHandlers.removeIf(eventHandler -> eventHandler.acceptor == listener);
    }

    @Override
    public void broadcast(OrchidEvent event) {
        for (Class<? extends OrchidEvent> inProgress : eventsInProgress) {
            if (event.getClass().equals(inProgress)) {
                throw new IllegalStateException(Clog.format("The event '#{$1}' is already in progress, it cannot be emitted again until this cycle has finished.", event.getClass().toString()));
            }
        }

        eventsInProgress.push(event.getClass());
        Clog.d("Broadcasting event: '#{$1}'", event.getClass().getSimpleName());
        for (EventHandler handler : this.eventHandlers) {
            try {
                callMethod(event, handler);
            }
            catch (Exception e) {
                throw new RuntimeException(e);
            }
        }
        eventsInProgress.pop();
    }

    private void callMethod(OrchidEvent event, EventHandler handler) throws Exception {
        Class<?> paramClass = handler.callback.getParameterTypes()[0];

        boolean callMethod = false;

        if(handler.allowSubclasses) {
            if(paramClass.isAssignableFrom(event.getClass())) {
                callMethod = true;
            }
        }
        else {
            if(paramClass.equals(event.getClass())) {
                callMethod = true;
            }
        }

        if(callMethod) {
            handler.callback.invoke(handler.acceptor, event);
        }
    }

}

File: OrchidCore/src/main/kotlin/com/eden/orchid/api/events/EventServiceImpl.kt
package com.eden.orchid.api.events

import com.caseyjbrooks.clog.Clog
import com.eden.orchid.api.OrchidContext
import com.eden.orchid.api.options.annotations.Description
import java.lang.reflect.Method
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
@Description(value = "The Orchid event broadcast system.", name = "Events")
@JvmSuppressWildcards
class EventServiceImpl
@Inject
constructor(
    private val eventListeners: Set<OrchidEventListener>
) : EventService {
    private lateinit var context: OrchidContext
    private val eventHandlers = mutableSetOf<EventHandler>()
    private val eventsInProgress = mutableListOf<Class<out OrchidEvent<*>>>()

    data class EventHandler(
        val eventClass: Class<out OrchidEvent<*>>,
        val allowSubclasses: Boolean = false,
        val acceptor: Any? = null,
        val callback: Method? = null
    )

    override fun initialize(context: OrchidContext) {
        this.context = context
        for (listener in eventListeners) {
            registerEventListeners(listener)
        }
    }

    override fun registerEventListeners(listener: OrchidEventListener) {
        listener.javaClass
            .declaredMethods
            .filter { method -> method.isAnnotationPresent(On::class.java) }
            .forEach { method ->
                val methodAnnotation = method.getAnnotation(On::class.java)!!
                val eventType = methodAnnotation.value.java
                val handler = EventHandler(
                    eventType,
                    methodAnnotation.subclasses,
                    listener,
                    method
                )

                var addHandler = true
                if (handler.callback!!.parameterCount != 1) {
                    addHandler = false
                    if (handler.allowSubclasses) {
                        Clog.e(
                            "Event handler [{}.{}] must accept a single parameter of type [{}] or a subclass of it.",
                            listener.javaClass.simpleName,
                            method.name,
                            eventType.simpleName
                        )
                    } else {
                        Clog.e(
                            "Event handler [{}.{}] must accept a single parameter of type [{}]",
                            listener.javaClass.simpleName,
                            method.name,
                            eventType.simpleName
                        )
                    }
                }
                if (handler.allowSubclasses) {
                    if (addHandler && !eventType.isAssignableFrom(handler.callback.parameterTypes[0])) {
                        addHandler = false
                        Clog.e(
                            "Event handler [{}.{}] must accept a single parameter of type [{}] or a subclass of it.",
                            listener.javaClass.simpleName,
                            method.name,
                            eventType.getSimpleName()
                        )
                    }
                } else {
                    if (addHandler && eventType != handler.callback.parameterTypes[0]) {
                        addHandler = false
                        Clog.e(
                            "Event handler [{}.{}] must accept a single parameter of type [{}]",
                            listener.javaClass.simpleName,
                            method.name,
                            eventType.getSimpleName()
                        )
                    }
                }
                if (addHandler) {
                    this.eventHandlers.add(handler)
                }
            }
    }

    override fun deregisterEventListeners(listener: OrchidEventListener) {
        eventHandlers.removeIf { eventHandler -> eventHandler.acceptor === listener }
    }

    @Synchronized
    override fun broadcast(event: OrchidEvent<*>) {
        for (inProgress in eventsInProgress) {
            if (event.javaClass == inProgress) {
                throw IllegalStateException(
                    Clog.format(
                        "The event \'#{$1}\' is already in progress, it cannot be emitted again until this cycle has finished.",
                        event.javaClass.toString()
                    )
                )
            }
        }
        eventsInProgress.add(event.javaClass)
        event.context = context

        for (handler in this.eventHandlers) {
            try {
                callMethod(event, handler)
            } catch (e: Exception) {
                throw RuntimeException(e)
            }

        }
        eventsInProgress.removeAt(eventsInProgress.size - 1)
    }

    @Throws(Exception::class)
    private fun callMethod(event: OrchidEvent<*>, handler: EventHandler) {
        val paramClass = handler.callback!!.parameterTypes[0]
        var callMethod = false
        if (handler.allowSubclasses) {
            if (paramClass.isAssignableFrom(event.javaClass)) {
                callMethod = true
            }
        } else {
            if (paramClass == event.javaClass) {
                callMethod = true
            }
        }
        if (callMethod) {
            handler.callback.invoke(handler.acceptor, event)
        }
    }

    fun getEventHandlers(): Set<EventHandler> {
        return this.eventHandlers
    }
}
